Add info about `inbox_event_root` on the header (#571)
FuelLabs/fuel-specs: src/protocol/relayer.md

# Layer 1 Relayer/Bridge Protocol

The Fuel relayer/bridge protocol is a set of rules that govern the interaction between the Fuel blockchain and the
Layer 1 (L1) blockchain (e.g. Ethereum).  

The Fuel blockchain can emit messages that will be processed by the smart contract on the L1 blockchain. The smart
contract on the L1 can also emit events that will be processed by the Fuel blockchain.
This is used to move any data between the L1 blockchain and the Fuel blockchain.

## Fuel Message Outbox

> The message outbox is the set of messages sent to the L1 blockchain from the Fuel blockchain.

Reviewer: I'd still like to see more technical description on how the message sending works. I.e. how can I verify that a message was actually sent on L1?

Author: I have intentionally not included a lot of detail on the "outbox" section, since this issue is specifically for the "inbox" stuff. There just wasn't any info on any of this stuff in the specs, so I wanted to set aside some space for the "outbox" stuff as well.

I'll create an issue for this: https://github.com/FuelLabs/fuel-specs/issues/575

1. Messages
2. Transactions

### Messages

An arbitrary message sent from the L1 to the Fuel blockchain. This can be used to move assets from the L1
to the Fuel blockchain or send other arbitrary information to the Fuel blockchain.

| name        | type    | description                                                         |
|-------------|---------|---------------------------------------------------------------------|
| `sender`    | `bytes[32]` | The identity of the sender of the message on the L1                 |
| `recipient` | `bytes[32]` | The recipient of the message on the Fuel Blockchain                 |
| `nonce`     | `bytes[32]` | Unique identifier of the message assigned by the L1 contract                                 |
| `amount`    | `uint64`  | The amount of the base asset transfer                              |
| `data`      | `byte[]`  | Arbitrary message data                                              |

### Transactions

These are transactions that are submitted on the L1 that must be executed on the Fuel blockchain.
This "Forced Transaction Inclusion" is a security feature that allows participants of the Fuel Blockchain to access

> their funds in the (unlikely) event that the Fuel blockchain block production is compromised or malicious.

Reviewer: Could mention censorship-resistancy here, as that's the primary technical form of compromise this protects against

| name                     | type      | description                                                                                                                                                    |
|--------------------------|-----------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `nonce`                  | `bytes[32]` | Unique identifier of the transaction assigned by the L1                                                                          contract                                                                    |
| `max_gas`                | `uint64`   | The maximum amount of gas allowed to use on  Fuel Blockchain                                                                          |

> | `serialized_transaction` | `byte[]`   | The serialized transaction bytes following canonical serialization. Only supports `Create` and `Script` [transaction](../tx-format/transaction.md), not `Mint` |

Reviewer 2: Now we have two more types of transactions=)

# Layer 1 Relayer/Bridge Protocol

The Fuel relayer/bridge protocol is a set of rules that govern the interaction between the Fuel blockchain and the
Layer 1 (L1) blockchain (e.g. Ethereum).  

The Fuel blockchain can emit messages that will be processed by the smart contract on the L1 blockchain. The smart
contract on the L1 can also emit events that will be processed by the Fuel blockchain.

> This is used to move assets between the two blockchains.

Reviewer 2: ```suggestion
This is used to move any data between L1 <> L2.
```

## Fuel Message Outbox

The message outbox is the set of messages sent to the L1 blockchain from the Fuel blockchain.

## Fuel Event Inbox

The event inbox is the set of events received from the L1 blockchain by the Fuel blockchain.

The block producer will receive the list of events from the L1 by some relayer, and then include both the count and the
merkle root of the events in the block header.

There are two types of events that can be received from the L1:

1. Messages
2. Transactions

### Messages

An arbitrary message sent from the L1 to the Fuel blockchain. This can be used to move assets from the L1
to the Fuel blockchain or send other arbitrary information to the Fuel blockchain.

| name        | type    | description                                                         |
|-------------|---------|---------------------------------------------------------------------|
| `sender`    | `bytes[32]` | The identity of the sender of the message on the L1                 |
| `recipient` | `bytes[32]` | The recipient of the message on the Fuel Blockchain                 |

> | `nonce`     | `bytes[32]` | Unique identifier of message on L1                                  |

Reviewer 2: ```suggestion
| `nonce`     | `bytes[32]` | Unique identifier of the message assigned by the L1 contract                                 |
```

| `amount`    | `uint64`  | The amount of the base asset transfer                              |
| `data`      | `byte[]`  | Arbitrary message data                                              |

> | `da_height` | `uint64` | The height of the L1 blockchain when this message event was emitted |

Reviewer 2: It would be nice to highlight that this fields is derived from Ethereum, not from the `MessageSent` event:

<img width="1086" alt="image" src="https://github.com/FuelLabs/fuel-specs/assets/18346821/a2cc790c-48c5-4a5e-b1f2-66a43eb1e759">

Author: Maybe we don't include it at all. We don't include this in the hash either:
![image](https://github.com/FuelLabs/fuel-specs/assets/11023047/fb141cd7-3a4d-4a8c-9083-901f5e520cfd)

### Transactions

These are transactions that are submitted on the L1 that must be executed on the Fuel blockchain.
This "Forced Transaction Inclusion" is a security feature that allows participants of the Fuel Blockchain to access

> there funds in the (unlikely) event that the Fuel blockchain block production is compromised or malicious.

Reviewer 2: ```suggestion
their funds in the (unlikely) event that the Fuel blockchain block production is compromised or malicious.
```

Reviewer 2: Maybe also mention that it is a way to bypass censoring by the network

| name                     | type      | description                                                                           |
|--------------------------|-----------|---------------------------------------------------------------------------------------|

> | `nonce`                  | `bytes[32]` | Unique identifier of the transaction on L1                                            |

Reviewer 2: ```suggestion
| `nonce`                  | `bytes[32]` | Unique identifier of the transaction assigned by the L1 contract                                                                    |
```

> | `max_gas`                | `uint64`   | The maximum amount of gas executing this transaction will cost on the Fuel Blockchain |

Reviewer 2: ```suggestion
| `max_gas`                | `uint64`   | The maximum amount of gas allowed to use on Fuel Blockchain |
```

> | `serialized_transaction` | `byte[]`   | The serialized transaction bytes                                                      |

Reviewer 2: Maybe put a link to canonical serialisation and to supported transaction types

> | `da_height`              | `uint64`   | The height of the L1 blockchain when this transaction event was emitted               |

Reviewer 2: The same here, would be nice to show that it is not a part of the event and derived from the Ethereum block.

Author: I'm just removing it. It's just a detail of our implementation of the specs, IMO.

# Layer 1 Relayer/Bridge Protocol

The Fuel relayer/bridge protocol is a set of rules that govern the interaction between the Fuel blockchain and the
Layer 1 (L1) blockchain (e.g. Ethereum).  

The Fuel blockchain can emit messages that will be processed by the smart contract on the L1 blockchain. The smart
contract on the L1 can also emit events that will be processed by the Fuel blockchain.
This is used to move any data between the L1 blockchain and the Fuel blockchain.

## Fuel Message Outbox

The message outbox is the set of messages sent to the L1 blockchain from the Fuel blockchain.

## Fuel Event Inbox

The event inbox is the set of events received from the L1 blockchain by the Fuel blockchain.

The block producer will receive the list of events from the L1 by some relayer, and then include both the count and the

> merkle root of the events in the block header.

Reviewer 3: It doesn't look like we actually put the event count into the header

Reviewer 3: This count is implicitly derived from the da height and reconciling that with L1 data

Author: Oh. Whoops. We only include it for the outbox.
